### ISSUE TITLE
equivalentId and canonicalId should be optional

### ISSUE THREAD

**Issue by peacekeeper** [w3c/did-core#717]
The spec currently says:

> The value of `equivalentId` _MUST_ be a set where each item is a string ...

> The value of `canonicalId` _MUST_ be a string that conforms to ...

I think this means that those metadata properties are currently required. I am pretty sure the intention of the Working Group was to make them optional and say something like the following instead:

> DID document metadata _MAY_ include a `equivalentId` property. If present, its value _MUST_ be a set where each item is a string ...

> DID document metadata _MAY_ include a `canonicalId` property. If present, its value _MUST_ be a string that conforms to ...

**Comment by csuwildcat** [comment]
@peacekeeper I think the wording is a bit confusing, but it is just saying that the values MUST be those type values IF the Method sets these properties. I do think your language is clearer on this point.

**Comment by TallTed** [comment]
Yeah, I think we had this wording a while back (for certain, on some properties, in some spec I was working on recently).  I would just change the two `The value of ...` to `If present, the value of ...` or maybe `If present, the value of optional property ...`

**Comment by peacekeeper** [comment]
My problem here is that all other metadata properties say very clearly if they SHOULD or MAY be included, whereas equivalentId and canonicalId only talk about their values, but are silent on whether they MUST/SHOULD/MAY be included in the metadata at all. So I think this needs to be clarified.

**Comment by TallTed** [comment]
@peacekeeper - Can you please add a link to the specific problem points in the doc, either in the source or the preview?  I think we're in agreement, that these two should be presented as optional in the same way as their neighbors are presented.  I (and, I believe, others) would prefer that the language used for these two match that already used for the others, whatever that has been, whether it's the complex phrasing you described above or the simpler phrasing I described...

**Comment by peacekeeper** [comment]
@TallTed for example, the section on `nextUpdate` says ([source here](https://github.com/w3c/did-core/blob/main/index.html#L3647) and [spec text here](https://www.w3.org/TR/2021/CR-did-core-20210318/#dfn-nextupdate)):

> DID document metadata MAY include a nextUpdate property if the resolved document version is not the latest version of the document.

On the other hand, the section on `equivalentId` says ([source here](https://github.com/w3c/did-core/blob/main/index.html#L3685) and [spec text here](https://www.w3.org/TR/2021/CR-did-core-20210318/#dfn-equivalentid)):

> The value of equivalentId MUST be a set where each item is a string that conforms to the rules in Section § 3.1 DID Syntax.

Does the rest of the section make it clear enough that this property is optional? I don't think so...

**Comment by TallTed** [comment]
@peacekeeper -- OK, I think I see your point. There's a lot more text about what `equivalentId` is for, both before and after the paragraph that describes what its expression MUST be, than there is about, for example, `nextUpdate`, and its optionality is more implied than explicitly stated within that extra text. It seems like a significant rewrite and restructure of those paragraphs on `equivalentId` and `canonicalId` will be necessary to move the basic structure and optionality to the start (as with the other properties), which will hopefully be quickly achievable and accepted as purely editorial. I look forward to your PR.

**Comment by csuwildcat** [comment]
The MUSTs should be:

1) if you are a Method that supports these properties, you MUST include any relevant canonical/equivalent values, per your Method
2) As a consumer, you MUST treat the canonical/equivalent values as defined and directed via the operative language of the definitions

**Comment by peacekeeper** [comment]
I created https://github.com/w3c/did-core/pull/720 to simply clarify that they are optional. Specific DID methods can add further instructions on how the two properties are used in those methods.

**Comment by csuwildcat** [comment]
They are optional for Methods to leverage, but required to be adhered to by
consumers of the DID Documents.

**Comment by TallTed** [comment]
[@csuwildcat]
> They are optional for Methods to leverage, but required to be adhered to by consumers of the DID Documents.

I do not understand what you are trying to say, that is not said by the existing text (after #720 is applied).  Perhaps you could make a rewording suggestion on #720?  Or make your own PR, if the relevant text isn't editable within #720?

**Comment by iherman** [comment]
The issue was discussed in a [meeting](https://www.w3.org/2019/did-wg/Meetings/Minutes/2021-03-30-did#section5-9) on 2021-03-30

- no resolutions were taken

<details><summary><i>View the transcript</i></summary>

#### 5.9. equivalentId and canonicalId should be optional 

_See github issue [#717](https://github.com/w3c/did-core/issues/717)._




**Brent Zundel:** Final issue. Assigned to Markus.  

**Markus Sabadello:** I came across this while I was working on my part of the test suite, implementing the resolution sections. I noticed that these two properties  
… unlike all the others, don't say whether they are optional or required.  
… Maybe me just misunderstanding ... It specifies what the values must be, but not if they are required or optional.  
… I'm sure the intention of the WG is to make them optional; raised a PR to clarify it.  
… Since sections marked at-risk, hope it does not cause complications.  

**Manu Sporny:** Reading the spec right now, I think it is a problem.  
… Trying to figure out if it pushes us into another CR phase to fix it.  
… It would be a normative change.  
… Could argue it was unclear whether optional or not. It didn't say you have to specify it.  
… We didn't do that, therefore it was optional. All we are doing is clarifying it is optional.  

> *Ted Thibodeau Jr.:* It's a change to a normative section, but I don't think it should force changed behavior

**Brent Zundel:** Comment said they are optional for methods to leverage.  
… Meaning the one implementer is intending them to be optional.  

**Drummond Reed:** I collaborated with Daniel on that text. It absolutely... I swear it has always been the intention.  
… I have a stack of bibles off-screen  

**Markus Sabadello:** Reading the actual text, it wasn't clear... The section marked at-risk: doesn't that mean we can make these changes?  

**Manu Sporny:** We didn't mark it at-risk in that way. We basically said we are seeking implementer feedback. If not enough, might remove ... not modify.  
… Grey area... As long as no one objections ... we just forgot to put the language there.  
… Unfortunately, intent doesn't matter. If it is possible someone looked at the specification and implemented it using just the words... If we are going to make a change to it, would it affect that person's implementation? I think good argument for no, since we didn't say you must have either of these properties.  

> *Kyle Den Hartog:* +1 to that interpretation

**Brent Zundel:** Agree  

</details>

**Comment by iherman** [comment]
The issue was discussed in a [meeting](https://www.w3.org/2019/did-wg/Meetings/Minutes/2021-04-06-did#section8-2) on 2021-04-06

- no resolutions were taken

<details><summary><i>View the transcript</i></summary>

#### 8.2. equivalentId and canonicalId should be optional 

_See github issue [#717](https://github.com/w3c/did-core/issues/717)._




_See github pull request [#720](https://github.com/w3c/did-core/pull/720)._




> *Daniel Buchner:* Let's fight!

**Daniel Burnett:** All involved people of this issue seem to be on the call, maybe we can move this forward.  

**Markus Sabadello:** I raised this because my reading of the spec text regarding the equivalent id and canonical id was that it felt like those are mandatory  
… or the spec did not say that they are optional  
… which is different from how other did doc metadata are described  
… I raised a simple PR that adds language to say that they are optional  
… I think that had support. Daniel then explained what this meant  
… and what did methods have to do  
… I'm not sure if the PR is sufficient or if there's anything else that needs to bed one  

**Daniel Buchner:** the interesting thing here is properties are required for methods that implement them. it's a directive that says if you're going to implement then you can't sometimes not throw them in  
… I don't know if we reflect that in the spec test.. optional in the general sense of did docs  
… if you choose to use these features it becomes required  
… not sure if we should put that in  

**Ted Thibodeau Jr.:** I'm fine with the text as its' in 720  
… it sounds like there's an additional piece  
… someplace else we said methods can be more stringent or can require something or can define things.. maybe there's another few words to throw in  
… I don't think this would count as a normative change to push another CR  
… the ambiguous test that suggested it was required would have meant that they forced it and the change to make it optional doesn't break anything  

**Daniel Buchner:** I can say that any method implementing these, the 4 or 5 today, would not be broken by changing it to optional in the general document sense  

**Ivan Herman:** daniel answered for those he knows about. Any change like that does it affect existing implementations?  
… if the answer is no then we are in a good place  

**Daniel Buchner:** the answer is no  
… if they didn't, if they were broken by this change it means they were misreading the rest of the properties  

**Daniel Burnett:** is there anything not said by the existing text after 720 is applied? take a look at 720 and if it's not complete make a suggestion or a new PR  
… thanks everyone  

> *Daniel Buchner:* will do

> *Daniel Buchner:* gtg, thank you for all your help (Justin, Dan, Markus, Ted, etc.)

---

</details>

**Comment by msporny** [comment]
PR #720 has been merged. Since both the issue and the PR were authored by the same individual, this issue is resolved assuming explicit confirmation that the PR addressed the author's concerns.
